Mathlib/Order/KonigLemma.lean

```diff
@@ -0,0 +1,147 @@
+/-
+Copyright (c) 2024 Peter Nelson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Peter Nelson
+-/
+import Mathlib.Order.Atoms.Finite
+import Mathlib.Order.Grade
+
+/-!
+# Kőnig's infinity lemma
+
+Kőnig's infinity lemma is most often stated as a graph theory result :
+every infinite, locally finite connected graph contains an infinite path.
+It has links to computability and proof theory, and
+is often applied when the graph is a tree whose vertices encode some other types of object.
+
+For applications of this form, it is convenient to state the lemma
+in terms of covers in a strongly atomic order rather than a graph;
+in this setting, the proof is almost trivial.
+In this file, we prove the lemma in this order-theoretic way,
+and also give a formulation that is convenient for applications in Ramsey theory.
+We defer the graph-theoretic version of the statement for future work.
```

Comment:
It would be worth explaining how this is a specialization of `nonempty_sections_of_finite_cofiltered_system`, but which has much lighter dependencies (that one goes through a category theory version of the compactness theorem, `TopCat.nonempty_limitCone_of_compact_t2_cofiltered_system`).

A less actionable suggestion: I think that this paragraph sort of apologizes for it not being the graph version, which doesn't seem necessary, since I think more often than not people use graphs as a way to organize the idea conceptually, vs the graph theoretic version being what people mean. I suppose I'd more likely think of it in terms of chains in a relation than order theory though, so that's worth explaining.

Regarding `exists_seq_forall_proj_of_forall_finite`, it seems like "main results" is justifying it as being useful for applications of Ramsey theory, but to me this sort of is what I think of when I think of Konig's lemma. It's the more category theoretic version. ("projective limits of nonempty finite types are nonempty"). Could you mention that this is the projective limit formulation?

Reply:
I wasn't aware of the category-theoretic connections. I have written something, but I don't follow your last paragraph, and I may have missed the point. If I have, can you suggest a phrasing?

Reply:
@kmill I ended up here while trying to figure out what is this reference to `nonempty_sections_of_finite_cofiltered_system`, which doesn't exist?

Reply:
I thought I remembered someone refactoring this, but it still seems to be around @digama0: https://leanprover-community.github.io/mathlib4_docs/Mathlib/CategoryTheory/CofilteredSystem.html#nonempty_sections_of_finite_cofiltered_system

Reply:
oops, I guess I messed up my search somehow.